### PR TITLE
Herwig7: Adding GoSam and update to newer Herwig7 version

### PR DESCRIPTION
**Pull request by mharrend**
In this PR besides updating of Herwig7 to Herwig7.0.4 from Herwig7.0.2,
also GoSam is added as a ME provider. This ME provider is necessary since it brings with itself the golem95 library which is also required by MG5aMC@NLO.

### CODE REVIEW

**Comment by mharrend** [comment]
Dear @smuzaffar @iahmad-khan,

could you please start the testing of this pr, too?

This an extension of #2707 which adds the GoSam event generator which is necessary since it brings the golem95 library within. This golem95 library is necessary for using any external ME provider with Herwig7-Matchbox interface.

Thanks, 

Marco

**Comment by smuzaffar** [comment]
please test

**Comment by mharrend** [comment]
Since Herwig7 needs the gosam_env.sh script which is not build via the Gosam tarball, 
GoSam is now installed via the python installation script.

However, there is a bug due to GoSam making use of the absolute Python path. This bug only occurs after a successful build of GoSam and during the installation.
@iahmad-khan Any ideas how one could fix this?

**Comment by iahmad-khan** [comment]
Do you mean python path is hard coded inside gosam ? becuse in the spec file , PYTHONPATH is set to new value before the install step.

**Comment by mharrend** [comment]
Yes, it seems that inside GoSam the python path is hardcoded or at least during the rpm installation the error suggests this.

**Comment by iahmad-khan** [comment]
i think you need to add the export keyword at the begging of the line that sets the PYTHONPATH

for example:

export PYTHONPATH=${CYTHON_ROOT}/${PYTHON_LIB_SITE_PACKAGES}

**Comment by mharrend** [comment]
The export PYTHONPATH=... part, unfortunately, had no positive effect.
I still get the following build report:

```
Building gosam. Log can be found in /afs/cern.ch/work/m/mharrend/cmsdist/buildneu/ext/CMSSW_9_0_X_2016-11-29-2300/20161206/r/BUILD/slc6_amd64_gcc530/external/gosam/2.0/log.
source /afs/cern.ch/work/m/mharrend/cmsdist/buildneu/ext/CMSSW_9_0_X_2016-11-29-2300/20161206/r/slc6_amd64_gcc530/rpm-env.sh ; TMPDIR=/afs/cern.ch/work/m/mharrend/cmsdist/buildneu/ext/CMSSW_9_0_X_2016-11-29-2300$
Build successful.
(<function buildPackage at 0xd7b938>, <Package name=gosam>, <scheduler.Scheduler object at 0xf53c90>) done
Trying to install the rpm package external+gosam+2.0 just built.
Checking local path dependency for rpm package external+gosam+2.0 just build.
Error: Local path dependency found for external+gosam+2.0
  /afs/cern.ch/work/m/mharrend/cmsdist/buildneu/ext/CMSSW_9_0_X_2016-11-29-2300/20161206/r/slc6_amd64_gcc530/external/python/2.7.11-oenich/bin/python
Requested to quit.
Requested to quit.
Requested to quit.
Requested to quit.
* The action "install-external+gosam+2.0" was not completed successfully because Traceback (most recent call last):
File "/afs/cern.ch/work/m/mharrend/cmsdist/buildneu/ext/CMSSW_9_0_X_2016-11-29-2300/20161206/PKGTOOLS/scheduler.py", line 199, in doSerial
result = commandSpec[0](*commandSpec[1:])
File "PKGTOOLS/cmsBuild", line 2946, in installPackage
RpmBuildFailed: Failed to build package gosam.

* The action "final-job" was not completed successfully because The following dependencies could not complete:
install-external+gosam+2.0
```

**Comment by iahmad-khan** [comment]
i think it needs to be build inside docker with base os image so that it can fail during the build. and then we will know what packages it tries to install from the system.

**Comment by davidlange6** [comment]
I’d suggest just grepping for the local path i the install attempt area. Most likely its in a bin area file..

**Comment by iahmad-khan** [comment]
Ok , yes i guess first i should try it that way.

**Comment by mharrend** [comment]
Sounds good to me. Please go ahead.
Thanks by the way

**Comment by iahmad-khan** [comment]
I build it locally but i couldnt figure out so for what is the local dependency it is complaning about.

```
Error: Local path dependency found for external+gosam+2.0-cms2
  /build/cmsbuild/newbuild/slc6_amd64_gcc530/external/python/2.7.11-oenich/bin/python

``` 

but the above line is not local to the system , it is a path from build area.

**Comment by iahmad-khan** [comment]
There are couple of things that @davidlt pointed out , first it should be build from source using the tarball ( by adding the tarball sources to a repo in cms-externals with some specific version ), and second , there are couple of dependencies that the install script is directly downloading such qgraf-3.1.4 , and form-4.1.033e etc that should,  instead , be added in requires list of gosam spec , from cms externals. ( by adding them to cms externals first)

**Comment by davidlange6** [comment]
The problem is the local build area dependency not the system one. Eg, 

/build/cmsbuild/newbuild…. needs to become /usr/bin/env python. There are numerous perl snippets in cmsdist to do that.

but it is basically:

perl -p -i -e "s|^#!.*python|#!/usr/bin/env python|" %{i}/bin/rootpy %{i}/bin/roosh %{i}/bin/root2hdf5

**Comment by mharrend** [comment]
@davidlange6 @iahmad-khan 
I believe I have found a solution.

I will switch back the gosam.spec file to its original version, so that the GoSam tarball together with a make, make install is used.
Afterwards, I will provide a /bin/gosam.py manually which is needed by Herwig7 to setup the GoSam environment.
It is a little bit ugly to write the env script by oneself, but I guess it is the best option.

What do you think?

**Comment by davidlt** [comment]
Yeah, the installer or something else generates some setup scripts which added full python paths into shebangs. Thus full path to Python is added to RPM requirements. I will just re-iterate what I already told to @iahmad-khan 
- Installed script downloads a tarball, but that's wrong as we cannot reproduce RPM with exact same content each time.
- Sadly configuration file has very specific version which maps to "latest" tarball, which is again not nice for reproducibility.
- There are extra dependencies being downloaded (3 or 4).

In order to make sure that we can reproduce our RPMs exactly in stable manner, we need to import the source code into cms-externals repositories, add additional dependencies as separate RPMs, etc.

**Comment by iahmad-khan** [comment]
@mharrend Finally i made a successfull build with this line added to `%install `  section of gosam.spec

     perl -p -i -e "s|^#!.*python|#!/usr/bin/env python$1|" $(grep -r -e "^#\!.*python.*" %i | cut -d: -f1)


So basically it finds all the python files and  replace the python interpreter path.

**Comment by mharrend** [comment]
Sorry, that it took me so long to come back to this issue.

I was now looking how we could include GoSam in a reproducible way. Unfortunately, I have no rights to open a new cms-externals repository.
But I would propose the following:

1) Add a GoSam repository to cms-externals
2) Add the GoSam install script from http://gosam.hepforge.org/gosam-installer/gosam_installer.py to the repository and modify the following line:
```
CONFIG_URL=r"https://gosam.hepforge.org/gosam-installer/gosam_install.cfg"
```
so that it points to a configuration file in the cms-externals repository.
3) Copy the configuration file  https://gosam.hepforge.org/gosam-installer/gosam_install.cfg to the cms-external repository and modify it.
4) Copy all external tarballs mentioned in the configuration file of 3) to the cms-external repository.
5) Make sure that the configuration file of 3) downloads the tarballs from 4) / the cms-external repository.

In this way, we could make use of the GoSam install script with slight modifications.

Looking on the many GoSam dependency I believe there is no good reason to build everything by itself.

What do you think?

If you could setup a GoSam cms-external repository, I could start to implement the steps mentioned above.

File: form.spec [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+
+CXX="$(which %{cms_cxx}) -fPIC"
+CC="$(which gcc) -fPIC"
+PLATF_CONF_OPTS="--enable-shared --disable-static"
+
+./configure $PLATF_CONF_OPTS \
+            --prefix=%i \
+            --bindir=%i/bin \
+            --without-gmp \
```
Comment (davidlt) [inline form.spec:23]: Why do you disable GMP, yet in `form-toolfile.spec` use say it's being used?

File: gmp.spec [hunk 2]
```diff
@@ -1,2 +1,2 @@
-### RPM external gmp 4.2.1
-Source: ftp://mirrors.kernel.org/gnu/%n/%n-%realversion.tar.gz
+### RPM external gmp 6.1.2
```
Comment (davidlt) [inline gmp.spec:1]: We don't use this SPEC. GMP is distributed as part of GCC toolchain package itself. The externals should use `gmp-static` package, which does not provide shared library, just static one.

File: gosam.spec [hunk 3]
```diff
@@ -1,36 +1,36 @@
...
+Requires: form
+Requires: gosamcontrib
 Requires: python cython
 
-%define keep_archives false
 
 %if "%{?cms_cxx:set}" != "set"
-%define cms_cxx c++
+%define cms_cxx g++
 %endif
```
Comment (davidlt) [inline gosam.spec]: You can remove this 4 line block and don't use `%cms_cxx`. It was never used.

File: gosamcontrib-toolfile.spec [hunk 4]
```diff
@@ -0,0 +1,23 @@
+### RPM external gosamcontrib-toolfile 1.0
+Requires: gosamcontrib
+%prep
+
+%build
+
+%install
+
+mkdir -p %i/etc/scram.d
+cat << \EOF_TOOLFILE >%i/etc/scram.d/form.xml
```
Comment (davidlt) [inline gosamcontrib-toolfile.spec]: This is wrong, looks to be copy & paste issue. This should be `gosamcontrib.xml`.

File: gosamcontrib.spec [hunk 5]
```diff
@@ -0,0 +1,35 @@
+### RPM external gosamcontrib 2.0-20150803
+Source: http://www.hepforge.org/archive/gosam/gosam-contrib-%{realversion}.tar.gz
+
+Requires: qgraf
+Requires: form
+
+
+%if "%{?cms_cxx:set}" != "set"
+%define cms_cxx g++
+%endif
```
Comment (davidlt) [inline gosamcontrib.spec]: Again, can be removed. We never used this once we put it.

File: gosamcontrib.spec [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+            --bindir=%i/bin \
+            --libdir=%i/lib \
+            CXX="$CXX" CC="$CC" FC="$FC" 
+
+make %makeprocesses  all
+
+%install
+make install 
+
+%post
```
Comment (davidlt) [inline gosamcontrib.spec]: You have nothing in the `%post` section thus remove it.

File: herwigpp.spec [hunk 7]
```diff
@@ -15,6 +15,7 @@ Requires: fastjet
 Requires: madgraph5amcatnlo
 Requires: gosam
 
+
```
Comment (davidlt) [inline herwigpp.spec]: Not needed.

File: herwigpp.spec [hunk 8]
```diff
@@ -52,8 +53,8 @@ PLATF_CONF_OPTS="--enable-shared --disable-static"
             --with-fastjet=$FASTJET_ROOT \
             --with-gsl=$GSL_ROOT \
             --with-boost=$BOOST_ROOT \
-	    --with-madgraph=$MADGRAPH5AMCATNLO_ROOT \
-	    --with-gosam=$GOSAM_ROOT \
+	          --with-madgraph=$MADGRAPH5AMCATNLO_ROOT \
```
Comment (davidlt) [inline herwigpp.spec]: Alignment issue.

File: qgraf.spec [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+${FC} qgraf*.f -o qgraf -O2
+
+
+
+%install
+mkdir %{i}/bin
+cp qgraf %{i}/bin
+
+
+%post
```
Comment (davidlt) [inline qgraf.spec]: Nothing in `%post` section thus remove it.

**Review by davidlt — CHANGES_REQUESTED** [review]
First quick review, more detailed is needed.

**Comment by mharrend** [comment]
@davidlange6 @smuzaffar @iahmad-khan 
The PR should now be finally ready.
Can you please start the testing?

File: gmp.spec (hunk 2 above)
Comment (mharrend) [reply gmp.spec:1]: It was part of the CMSDIST and I got a compiler error when I was building it. So I decided to update it and the build error went away.

File: form.spec (hunk 1 above)
Comment (mharrend) [reply form.spec:23]: I first tried it with gmp but always the libgmp.so.3 was not, thus, I removed it in the form.spec and forgot to do it in the form-toolfile.spec, too.

**Comment by mharrend** [comment]
@davidlt Can you review it again, please?

Thanks

**Comment by mharrend** [comment]
@davidlt @davidlange6 I have commented all the reviewed points and also changed it according to your proposal. So, can you please start the tests or comment further?

Thanks, Marco

**Comment by smuzaffar** [comment]
@iahmad-khan , can you please try building it locally?

File: gosam.spec [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+
+
+
+%prep
+%setup -q -n gosam
+
+
+
+%build
+CXX="$(which %{cms_cxx}) -fPIC"
```
Comment (smuzaffar) [inline gosam.spec]: cms_cxx is not defined. I would suggest to replace it directly with g++. same for gosamcontrib.spec

**Comment by smuzaffar** [comment]
@mharrend, @iahmad-khan is working on it and will make a PR for 91X

**Comment by mharrend** [comment]
@smuzaffar @iahmad-khan Thanks for working on it! I just updated the cms_cxx part. But will now refrain from updating parts.

**Comment by smuzaffar** [comment]
closing it in favor of #2889 for 91X

**Comment by Andrej-CMS** [comment]
Hello @mharrend, @smuzaffar , @iahmad-khan , @perrozzi,

Would you please explain why this PR was closed in favor of 91X? These changes are necessary for Herwig7 in order to work in CMSSW_9_0_X. 

Cheers,
Andrej

**Comment by smuzaffar** [comment]
@Andrej-CMS , new externals should first go in dev release (which is 91X now). There are also some fixes in  #2889 too (which are missing here). Once that PR is in 91X (and if needed for 90X) then we will create a backport PR for 90X.

**Comment by Andrej-CMS** [comment]
@smuzaffar thank you very much for the fast reply! I've seen that #2889 ist on hold for quite some time. Will there be a follow up on the PR in the near future? 

As I already said, this is required for the new Herwig7 interface for CMSSW

**Comment by smuzaffar** [comment]
I guess there might be license related issue as gosam and form are GNU GPL v3. I am not a guru on license so I let @davidlt and/or @davidlange6 to comment

**Pull request closed without merging**